File: packages/image-theme-colors/android/src/main/java/expo/modules/imagethemecolors/ExpoImageThemeColorsModule.kt
@file:OptIn(EitherType::class)

package expo.modules.imagethemecolors

import android.graphics.Bitmap
import android.graphics.Bitmap.Config
import android.graphics.Canvas
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import androidx.palette.graphics.Palette
import expo.modules.kotlin.apifeatures.EitherType
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.toCodedException
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.sharedobjects.SharedRef
import expo.modules.kotlin.types.EitherOfThree
import expo.modules.kotlin.types.OptimizedRecord
import expo.modules.kotlin.types.toKClass

internal class ImageLoadingFailedException(cause: CodedException?) :
    CodedException(message = "Could not load the image from sharedRef", cause)

@OptimizedRecord
class SwatchRecord : Record {
    @Field
    var hex: String = ""

    @Field
    var titleTextColor: String = ""

    @Field
    var bodyTextColor: String = ""

    @Field
    var population: Int = 0
}

@OptimizedRecord
class PaletteRecord : Record {
    @Field
    var width: Int = 0

    @Field
    var height: Int = 0

    @Field
    var dominant: SwatchRecord? = null

    @Field
    var vibrant: SwatchRecord? = null

    @Field
    var lightVibrant: SwatchRecord? = null

    @Field
    var darkVibrant: SwatchRecord? = null

    @Field
    var muted: SwatchRecord? = null

    @Field
    var lightMuted: SwatchRecord? = null

    @Field
    var darkMuted: SwatchRecord? = null
}

class ExpoImageThemeColorsModule : Module() {
    companion object {
        private const val TAG = "ExpoImageThemeColor"
    }

    override fun definition() = ModuleDefinition {
        Name("ExpoImageThemeColors")

        AsyncFunction("extractThemeColorAsync") Coroutine { imageSource: EitherOfThree<String, SharedRef<Bitmap>, SharedRef<Drawable>>
            ->
            val bitmap = when {
                imageSource.`is`(String::class) -> getBitmapFromUrl(imageSource.get(String::class))
                imageSource.`is`(toKClass<SharedRef<Bitmap>>()) -> imageSource.get(toKClass<SharedRef<Bitmap>>()).ref
                else -> drawableToBitmap(imageSource.get(toKClass<SharedRef<Drawable>>()).ref)
            }
            android.util.Log.d(TAG, "get bitmap")

            val palette = Palette.from(bitmap).generate()

            return@Coroutine PaletteRecord().apply {
                width = bitmap.width
                height = bitmap.height
                dominant = palette.dominantSwatch.toSwatchRecord()
                vibrant = palette.vibrantSwatch.toSwatchRecord()
                lightVibrant = palette.lightVibrantSwatch.toSwatchRecord()
                darkVibrant = palette.darkVibrantSwatch.toSwatchRecord()
                muted = palette.mutedSwatch.toSwatchRecord()
                lightMuted = palette.lightMutedSwatch.toSwatchRecord()
                darkMuted = palette.darkMutedSwatch.toSwatchRecord()
            }
        }
    }

    private fun getBitmapFromUrl(urlString: String): Bitmap {
        try {
            val url = java.net.URL(urlString)
            return android.graphics.BitmapFactory.decodeStream(url.openStream())
        } catch (e: Exception) {
            throw ImageLoadingFailedException(e.toCodedException())
        }
    }

    private fun drawableToBitmap(drawable: Drawable): Bitmap {
        try {
            if (drawable is BitmapDrawable) return drawable.bitmap

            val width = drawable.intrinsicWidth.coerceAtLeast(1)
            val height = drawable.intrinsicHeight.coerceAtLeast(1)
            val bitmap = Bitmap.createBitmap(width, height, Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            drawable.setBounds(0, 0, canvas.width, canvas.height)
            drawable.draw(canvas)
            return bitmap
        } catch (e: Exception) {
            throw CodedException(
                message = "Failed to convert drawable to bitmap: ${e.message}",
                cause = null
            )
        }
    }


    private fun Int.toHexColor(): String {
        return String.format("#%06X", (0xFFFFFF and this))
    }

    private fun Palette.Swatch?.toSwatchRecord(): SwatchRecord? {
        if (this == null) {
            return null
        }

        val swatch = this
        return SwatchRecord().apply {
            hex = swatch.rgb.toHexColor()
            titleTextColor = swatch.titleTextColor.toHexColor()
            bodyTextColor = swatch.bodyTextColor.toHexColor()
            population = swatch.population
        }
    }
}

File: packages/image-theme-colors/ios/ExpoImageThemeColorsModule.swift
import ExpoModulesCore
import swift_vibrant
import UIKit

struct SwatchRecord: Record {
  @Field
  var hex: String = ""

  @Field
  var titleTextColor: String = ""

  @Field
  var bodyTextColor: String = ""

  @Field
  var population: Int = 0
}

struct PaletteRecord: Record {
  @Field
  var width: Double = 0

  @Field
  var height: Double = 0

  @Field
  var dominant: SwatchRecord?

  @Field
  var vibrant: SwatchRecord?

  @Field
  var lightVibrant: SwatchRecord?

  @Field
  var darkVibrant: SwatchRecord?

  @Field
  var muted: SwatchRecord?

  @Field
  var lightMuted: SwatchRecord?

  @Field
  var darkMuted: SwatchRecord?
}

public class ExpoImageThemeColorsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoImageThemeColors")

    AsyncFunction("extractThemeColorAsync") { (source: Either<URL, SharedRef<UIImage>>) -> PaletteRecord in
        let image: UIImage
        
        if let url: URL = source.get() {
            // Load image from URL
            let data = try Data(contentsOf: url)
            guard let img = UIImage(data: data) else {
                throw Exception(name: "ImageLoadingFailed", description: "Could not load image from URL")
            }
            image = img
        } else if let sharedRef: SharedRef<UIImage> = source.get() {
            image = sharedRef.ref
        } else {
             throw Exception(name: "InvalidSource", description: "Invalid image source provided")
        }

        // Generate palette
        let palette = Vibrant.from(image).getPalette()
        
        let record = PaletteRecord()
        record.width = Double(image.size.width)
        record.height = Double(image.size.height)
        record.dominant = (palette.Vibrant ?? palette.Muted)?.toRecord()
        record.vibrant = palette.Vibrant?.toRecord()
        record.lightVibrant = palette.LightVibrant?.toRecord()
        record.darkVibrant = palette.DarkVibrant?.toRecord()
        record.muted = palette.Muted?.toRecord()
        record.lightMuted = palette.LightMuted?.toRecord()
        record.darkMuted = palette.DarkMuted?.toRecord()

        return record
    }
  }
}

extension Swatch {
    func toRecord() -> SwatchRecord {
        let record = SwatchRecord()
        record.hex = self.uiColor.toHexString()
        record.titleTextColor = self.titleTextColor.toHexString()
        record.bodyTextColor = self.bodyTextColor.toHexString()
        record.population = self.population
        return record
    }
}

extension UIColor {
    func toHexString() -> String {
        var r: CGFloat = 0
        var g: CGFloat = 0
        var b: CGFloat = 0
        var a: CGFloat = 0
        // Use getRed to handle different color spaces (like Display P3)
        if self.getRed(&r, green: &g, blue: &b, alpha: &a) {
            let rgb: Int = (Int)(r*255)<<16 | (Int)(g*255)<<8 | (Int)(b*255)<<0
            return String(format:"#%06X", rgb)
        }
        return "#000000"
    }
}
